File: go/pascals-triangle/pascals_triangle.go
package pascal

// Triangle computes Pascal's triangle up to a given number of rows
func Triangle(n int) (triangle [][]int) {
	if n < 1 {
		return
	}

	triangle = append(triangle, []int{1})

	for i := 2; i <= n; i++ {
		last := triangle[len(triangle)-1]
		lastLen := len(last)

		row := make([]int, lastLen+1)

		row[0] = last[0]

		for j := 0; j < lastLen-1; j++ {
			row[j+1] = last[j] + last[j+1]
		}

		row[len(row)-1] = last[lastLen-1]

		triangle = append(triangle, row)
	}

	return
}

File: ruby/pascals-triangle/pascals_triangle.rb
class Triangle
  def initialize(height)
    @height = height
  end

  def rows
    rows = [[1]]

    (@height - 1).times do
      row = []
      last_row = rows.last

      (last_row.length - 1).times do |i|
        row << last_row[i] + last_row[i+1]
      end

      row.unshift(last_row.first)
      row << last_row.last

      rows << row
    end

    rows
  end
end
